// hastily/build/module/lib/mappers/resize.js
export const optoFitToSharp = {
    bounds: 'inside',
    cover: 'outside',
    crop: 'cover',
};
// sharp supports mitchell interpolation and not bilinear,
// but who's gonna notice?
export const optoResizeFilterToSharp = {
    bicubic: 'cubic',
    bilinear: 'mitchell',
    cubic: 'cubic',
    lanczos: 'lanczos3',
    lanczos2: 'lanczos2',
    lanczos3: 'lanczos3',
    linear: 'mitchell',
    nearest: 'nearest',
};
/**
 * @hidden
 */
const resize = (sharp, params) => {
    if (!params.has('width')) {
        params.warn('unsupported', 'width', 'resize without width param. Will not resize.');
        return false;
    }
    const nums = params.toNumbers(['width', 'height', 'dpr']);
    const width = nums[0];
    let height = nums[1];
    const dpr = nums[2];
    for (const [name, param] of [
        ['width', width],
        ['height', height],
    ]) {
        if (param < 1 && param > 0) {
            params.warn('unsupported', name, 'ratio-based resize/crop. Will not apply.');
            return false;
        }
    }
    if (isNaN(width)) {
        params.warn('invalid', 'width', 'resize with non-numeric width param');
        return false;
    }
    if (isNaN(height)) {
        height = undefined;
    }
    const { log } = params;
    log.debug('width %s, height %s, dpr %s', width, height, dpr);
    const options = {
        withoutEnlargement: true,
    };
    if (params.has('fit')) {
        const fitOpt = params.get('fit');
        const fit = optoFitToSharp[fitOpt];
        if (fit) {
            options.fit = fit;
        }
        else {
            params.warn('unsupported', 'fit');
        }
        log.debug('mapped options.fit from %s -> %s', fitOpt, options.fit);
    }
    else if (typeof height === 'number' && height > 0) {
        options.fit = 'fill';
    }
    if (params.get('enable') === 'true' || params.get('disable') === 'false') {
        log.debug('overriding withoutEnlargement to false, yucky enlargement enabled');
        options.withoutEnlargement = false;
    }
    if (params.has('resize-filter')) {
        const filterOpt = params.get('resize-filter');
        const kernel = optoResizeFilterToSharp[filterOpt];
        if (kernel) {
            options.kernel = kernel;
        }
        else {
            params.warn('unsupported', 'fit');
        }
        log.debug('mapped options.resize-filter from %s -> %s', filterOpt, options.kernel);
    }
    log.debug('sharp.resize(%s, %s, %o)', width, height, options);
    return sharp.resize(width, height, options);
};
export default resize;
