// pertain/build/module/ExplicitDependencySet.js
import ExplicitDependency from './ExplicitDependency';
import TopologicalSorter from './TopologicalSorter';
/**
 * A set of modules that can be queried for pertaining to a given subject.
 */
export default class ExplicitDependencySet {
    resolve;
    dependencies;
    /**
     * Cache storage to prevent unnecessary recalculation of the same list.
     * The same subject should result in the same list while this object exists.
     */
    sortedBySubject;
    constructor(resolve, names) {
        this.resolve = resolve;
        this.dependencies = [];
        this.sortedBySubject = new Map();
        names.forEach((name) => this.add(name));
    }
    /**
     * Gets a list of ExplicitDependencies in this set which pertain to the
     * subject (that is, their `package.json` has a valid key for the subject
     * that indicates a requireable file). Detects dependencies between the
     * packages that pertain and sorts the list in dependency order.
     */
    pertaining(subject) {
        let sorted = this.sortedBySubject.get(subject);
        if (!sorted) {
            const pertaining = this.dependencies.filter((dependency) => dependency.pertains(subject));
            // Returns a list of dependents (not dependencies) of the supplied
            // dependency. This is the data structure needed for an efficient
            // topological sort.
            const getOutgoingEdges = (dependency) => pertaining.filter((dependent) => dependent !== dependency && dependent.dependsOn(dependency.name));
            const sorter = new TopologicalSorter(getOutgoingEdges);
            sorted = sorter.sort(pertaining);
            this.sortedBySubject.set(subject, sorted);
        }
        return sorted;
    }
    /**
     * Only add dependencies which can be resolved in the first place. If there
     * is a problem finding them, just skip them; they don't pertain!
     */
    add(name) {
        const modulePath = this.resolve(name);
        if (modulePath) {
            const dependency = new ExplicitDependency(modulePath);
            this.dependencies.push(dependency);
        }
    }
}
