// pertain/build/module/PackageJson.js
import objectPath from 'object-path';
import { readFileSync } from 'fs';
import path from 'path';
/**
 * Lazy-loading proxy for package.json. A subset of properties are available,
 * and a static method
 * `PackageJson.lookup(package: PackageJson, dotPath: string)`
 * exists for looking up other arbitrary properties.
 *
 * This speeds up dependency gathering; it won't be necessary to read all
 * package files from disk.
 */
export default class PackageJson {
    json;
    lookup(dotPath, fallback) {
        return this.getJson().get(dotPath) || fallback;
    }
    get dependencies() {
        return this.lookup('dependencies', {});
    }
    get devDependencies() {
        return this.lookup('devDependencies', {});
    }
    get name() {
        return this.lookup('name');
    }
    get peerDependencies() {
        return this.lookup('peerDependencies', {});
    }
    modulePath;
    constructor(modulePath) {
        this.modulePath = modulePath;
    }
    getJson() {
        if (!this.json) {
            this.json = objectPath(JSON.parse(readFileSync(path.join(this.modulePath, 'package.json'), 'utf8')));
        }
        return this.json;
    }
}
