// pertain/build/main/ExplicitDependencySet.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ExplicitDependency_1 = __importDefault(require("./ExplicitDependency"));
const TopologicalSorter_1 = __importDefault(require("./TopologicalSorter"));
/**
 * A set of modules that can be queried for pertaining to a given subject.
 */
class ExplicitDependencySet {
    constructor(resolve, names) {
        this.resolve = resolve;
        this.dependencies = [];
        this.sortedBySubject = new Map();
        names.forEach((name) => this.add(name));
    }
    /**
     * Gets a list of ExplicitDependencies in this set which pertain to the
     * subject (that is, their `package.json` has a valid key for the subject
     * that indicates a requireable file). Detects dependencies between the
     * packages that pertain and sorts the list in dependency order.
     */
    pertaining(subject) {
        let sorted = this.sortedBySubject.get(subject);
        if (!sorted) {
            const pertaining = this.dependencies.filter((dependency) => dependency.pertains(subject));
            // Returns a list of dependents (not dependencies) of the supplied
            // dependency. This is the data structure needed for an efficient
            // topological sort.
            const getOutgoingEdges = (dependency) => pertaining.filter((dependent) => dependent !== dependency && dependent.dependsOn(dependency.name));
            const sorter = new TopologicalSorter_1.default(getOutgoingEdges);
            sorted = sorter.sort(pertaining);
            this.sortedBySubject.set(subject, sorted);
        }
        return sorted;
    }
    /**
     * Only add dependencies which can be resolved in the first place. If there
     * is a problem finding them, just skip them; they don't pertain!
     */
    add(name) {
        const modulePath = this.resolve(name);
        if (modulePath) {
            const dependency = new ExplicitDependency_1.default(modulePath);
            this.dependencies.push(dependency);
        }
    }
}
exports.default = ExplicitDependencySet;
