// pertain/build/main/PackageJson.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const object_path_1 = __importDefault(require("object-path"));
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
/**
 * Lazy-loading proxy for package.json. A subset of properties are available,
 * and a static method
 * `PackageJson.lookup(package: PackageJson, dotPath: string)`
 * exists for looking up other arbitrary properties.
 *
 * This speeds up dependency gathering; it won't be necessary to read all
 * package files from disk.
 */
class PackageJson {
    lookup(dotPath, fallback) {
        return this.getJson().get(dotPath) || fallback;
    }
    get dependencies() {
        return this.lookup('dependencies', {});
    }
    get devDependencies() {
        return this.lookup('devDependencies', {});
    }
    get name() {
        return this.lookup('name');
    }
    get peerDependencies() {
        return this.lookup('peerDependencies', {});
    }
    constructor(modulePath) {
        this.modulePath = modulePath;
    }
    getJson() {
        if (!this.json) {
            this.json = (0, object_path_1.default)(JSON.parse((0, fs_1.readFileSync)(path_1.default.join(this.modulePath, 'package.json'), 'utf8')));
        }
        return this.json;
    }
}
exports.default = PackageJson;
