// hastily/build/main/lib/mappers/__testhelpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runMapperWithParams = exports.mockParams = void 0;
const stream_1 = require("stream");
const url_1 = require("url");
const fastly_params_1 = __importDefault(require("../fastly-params"));
class MockSharp extends stream_1.Duplex {
    constructor() {
        super(...arguments);
        this.calls = [];
    }
    removeAlpha() {
        throw new Error('Method not implemented.');
    }
    ensureAlpha() {
        throw new Error('Method not implemented.');
    }
    extractChannel(..._) {
        throw new Error('Method not implemented.');
    }
    joinChannel(..._) {
        throw new Error('Method not implemented.');
    }
    bandbool(..._) {
        throw new Error('Method not implemented.');
    }
    tint(..._) {
        throw new Error('Method not implemented.');
    }
    greyscale(..._) {
        throw new Error('Method not implemented.');
    }
    grayscale(..._) {
        throw new Error('Method not implemented.');
    }
    toColourspace(..._) {
        throw new Error('Method not implemented.');
    }
    toColorspace(..._) {
        throw new Error('Method not implemented.');
    }
    composite(..._) {
        throw new Error('Method not implemented.');
    }
    clone() {
        throw new Error('Method not implemented.');
    }
    metadata(..._) {
        throw new Error('Method not implemented.');
    }
    stats(..._) {
        throw new Error('Method not implemented.');
    }
    limitInputPixels(..._) {
        throw new Error('Method not implemented.');
    }
    sequentialRead(..._) {
        throw new Error('Method not implemented.');
    }
    rotate(_angle, _options) {
        return this.track('rotate', arguments);
    }
    flip(_flip) {
        return this.track('flip', arguments);
    }
    flop(_flop) {
        return this.track('flop', arguments);
    }
    sharpen(_sigma, _flat, _jagged) {
        return this.track('sharpen', arguments);
    }
    median(_size) {
        return this.track('median', arguments);
    }
    blur(_sigma) {
        return this.track('blur', arguments);
    }
    flatten(_flatten) {
        return this.track('flatten', arguments);
    }
    gamma(_gamma) {
        return this.track('gamma', arguments);
    }
    negate(_negate) {
        return this.track('negate', arguments);
    }
    normalise(_normalise) {
        return this.track('normalise', arguments);
    }
    normalize(_normalize) {
        return this.track('normalize', arguments);
    }
    convolve(_kernel) {
        return this.track('convolve', arguments);
    }
    threshold(_threshold, _options) {
        return this.track('threshold', arguments);
    }
    boolean(_operand, _operator, _options) {
        return this.track('boolean', arguments);
    }
    linear(_a, _b) {
        return this.track('linear', arguments);
    }
    recomb(_matrix) {
        return this.track('recomb', arguments);
    }
    // public recomb(
    //   _inputMatrix: [
    //     [number, number, number],
    //     [number, number, number],
    //     [number, number, number]
    //   ]
    // ) {
    //   return this.track('recomb', arguments);
    // }
    modulate(..._) {
        throw new Error('Method not implemented.');
    }
    toFile(..._) {
        throw new Error('Method not implemented.');
    }
    toBuffer(options) {
        if (typeof options === 'function') {
            setImmediate(() => options(new Error(), Buffer.from([]), {}));
            return this;
        }
        if (options === undefined) {
            return Promise.resolve(Buffer.from([]));
        }
        const resolved = {
            data: Buffer.from([]),
            info: {
                channels: 2,
                format: 'nothing',
                height: 10,
                premultiplied: false,
                size: 10,
                width: 10,
            },
        };
        return Promise.resolve(resolved);
    }
    withMetadata(..._) {
        throw new Error('Method not implemented.');
    }
    jpeg(..._) {
        throw new Error('Method not implemented.');
    }
    png(..._) {
        throw new Error('Method not implemented.');
    }
    webp(..._) {
        throw new Error('Method not implemented.');
    }
    tiff(..._) {
        throw new Error('Method not implemented.');
    }
    raw() {
        throw new Error('Method not implemented.');
    }
    toFormat(_format, _options) {
        return this.track('toFormat', arguments);
    }
    tile(_tile) {
        return this.track('tile', arguments);
    }
    resize(_width, _height, _options) {
        return this.track('resize', arguments);
    }
    extend(_extend) {
        return this.track('extend', arguments);
    }
    extract(_region) {
        return this.track('extract', arguments);
    }
    trim(_threshold) {
        return this.track('threshold', arguments);
    }
    track(name, args) {
        this.calls.push([name, [...args]]);
        return this;
    }
}
function mockParams(query) {
    return new fastly_params_1.default(new Map(new url_1.URLSearchParams(query).entries()), {}, {});
}
exports.mockParams = mockParams;
function runMapperWithParams(mapper, query) {
    const mock = new MockSharp();
    const params = mockParams(query);
    return {
        mapped: mapper(mock, params),
        mock,
        warnings: params.getWarnings(),
    };
}
exports.runMapperWithParams = runMapperWithParams;
