// hastily/build/main/lib/imageopto.js
"use strict";
/**
 * Simulate the [Fastly ImageOpto](https://docs.fastly.com/api/imageopto/) API.
 * Returns a middleware which attempts to run image responses through `sharp`.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.imageopto = exports.hasSupportedExtension = exports.HASTILY_STREAMABLE_PATH_REGEXP = exports.HASTILY_STREAMABLE_FILETYPES = exports.HASTILY_HEADER = void 0;
const logging_1 = require("./logging");
const mime_types_1 = __importDefault(require("mime-types"));
const sharp_1 = __importDefault(require("sharp"));
const url_1 = require("url");
const vary_1 = __importDefault(require("vary"));
const errors_1 = __importDefault(require("./errors"));
const fastly_params_1 = __importDefault(require("./fastly-params"));
const map_options_1 = __importDefault(require("./map-options"));
const splice_response_1 = __importDefault(require("./splice-response"));
exports.HASTILY_HEADER = {
    NAME: 'X-Optimized',
    VALUE: 'hastily',
};
exports.HASTILY_STREAMABLE_FILETYPES = new Set(Object.keys(sharp_1.default.format).filter((ext) => sharp_1.default.format[ext].input.stream));
// plus the one sharp doesn't mention outright, the alias to jpeg...
exports.HASTILY_STREAMABLE_FILETYPES.add('jpg');
// minus SVGs, which are widely supported in 2019 and we should not rasterize
exports.HASTILY_STREAMABLE_FILETYPES.delete('svg');
exports.HASTILY_STREAMABLE_PATH_REGEXP = new RegExp(`/.+\\.(${[...exports.HASTILY_STREAMABLE_FILETYPES].join('|')})(?:[?#].*)?`);
/**
 * Use the `sharp.format` manifest to determine if the current request's file
 * extension matches a format that sharp can stream in to optimize.
 * @param req {Request}
 */
exports.hasSupportedExtension = (req) => exports.HASTILY_STREAMABLE_PATH_REGEXP.test(req.originalUrl);
/**
 * Returns a new imageopto middleware for use in Express `app.use()`.
 * Won't do anything if the Express app isn't already serving images!
 *
 */
function imageopto(filterOrOpts) {
    const options = {
        filter: exports.hasSupportedExtension,
        force: false,
    };
    if (typeof filterOrOpts === 'function') {
        options.filter = filterOrOpts;
    }
    else if (typeof filterOrOpts === 'object') {
        options.filter = filterOrOpts.filter || options.filter;
        options.force = filterOrOpts.force;
        options.errorLog = options.quiet ? (_) => void 0 : filterOrOpts.errorLog;
    }
    const constructorLog = logging_1.createLogger('middleware');
    constructorLog.debug('creating new middleware');
    if (options.filter === exports.hasSupportedExtension) {
        constructorLog.debug('middleware filtering req.path for %s', exports.HASTILY_STREAMABLE_PATH_REGEXP.source);
    }
    const requestLog = logging_1.createLogger('request');
    return function hastily(req, res, next) {
        const reqLog = requestLog.child({ req });
        if (!options.filter(req)) {
            reqLog.debug('did not pass supplied filter function');
            return next();
        }
        splice_response_1.default(req, res, next, () => {
            // determine if the entity should be transformed
            if (!shouldTransform(req, res, reqLog, options)) {
                return false;
            }
            reqLog.debug('hastily will handle this image by transforming the response through sharp');
            vary_1.default(res, 'Accept');
            res.setHeader(exports.HASTILY_HEADER.NAME, exports.HASTILY_HEADER.VALUE);
            // image opto stream
            const params = new fastly_params_1.default(new Map(new url_1.URLSearchParams(req.query).entries()), req, res);
            const sharpStream = sharp_1.default();
            const emitSharpError = (error) => {
                if (options.errorLog) {
                    options.errorLog(error);
                }
                else {
                    reqLog.error('Image processing failed: %s', error.toString());
                }
            };
            sharpStream.on('error', emitSharpError);
            try {
                const transformStream = map_options_1.default(params, sharpStream);
                const warnings = params.getWarnings();
                if (warnings.length > 0) {
                    const requestErrors = new errors_1.default(req.url, warnings);
                    if (options.errorLog) {
                        options.errorLog(requestErrors);
                    }
                    else {
                        reqLog.warn(requestErrors.toString());
                    }
                }
                reqLog.debug('mapped options and created sharp stream');
                return transformStream;
            }
            catch (e) {
                emitSharpError(e);
                return false;
            }
        });
    };
}
exports.imageopto = imageopto;
const cacheControlNoTransformRegExp = /(?:^|,)\s*?no-transform\s*?(?:,|$)/;
/**
 * Determine if the entity should be transformed.
 * @private
 */
function shouldTransform(req, res, reqLog, options) {
    if (req.method !== 'GET') {
        reqLog.debug('no transform: request method must be GET but is %s', req.method);
        return false;
    }
    if (res.statusCode > 299 || res.statusCode < 200) {
        reqLog.debug('no transform: res.statusCode must be 2xx but is %s', res.statusCode);
        return false;
    }
    const cacheControl = res.getHeader('Cache-Control');
    // Don't optimize for Cache-Control: no-transform
    // https://tools.ietf.org/html/rfc7234#section-5.2.2.4
    if (cacheControl &&
        cacheControlNoTransformRegExp.test(cacheControl.toString())) {
        reqLog.debug('no transform: cache control header: "%s"', cacheControl);
        return false;
    }
    if (!options.force) {
        // Don't optimize if we've already done it somewhere
        const hastilyHeader = res.getHeader(exports.HASTILY_HEADER.NAME);
        if (hastilyHeader === exports.HASTILY_HEADER.VALUE) {
            reqLog.warn('no transform: header %o, hastily alrady transformed this earlier', exports.HASTILY_HEADER);
            return false;
        }
        // Don't optimize if Fastly has already done it for us
        const fastlyHeader = res.getHeader('fastly-io-info');
        if (fastlyHeader) {
            reqLog.debug('no transform: fastly already transformed according to fastly-io-info header: "%s"', fastlyHeader);
            return false;
        }
    }
    const contentType = res.getHeader('content-type');
    const extension = mime_types_1.default.extension(contentType);
    if (!extension) {
        reqLog.error('no transform: no valid content-type could not be detected');
        return false;
    }
    const sharpFormatCapabilities = sharp_1.default.format[extension];
    if (!sharpFormatCapabilities || !sharpFormatCapabilities.input.stream) {
        reqLog.error('no transform: sharp does not support input of type "%s"', contentType);
        return false;
    }
    const contentEncoding = res.getHeader('content-encoding');
    if (contentEncoding &&
        typeof contentEncoding === 'string' &&
        contentEncoding.trim()) {
        reqLog.error('no transform: image is compressed with content-encoding "%s"; hastily does not support decompressing images. The server *should not* be compressing images! HTTP content encoding is only recommended for text documents. Binary files will not get smaller, and may evem get larger!');
    }
    return true;
}
