// hastily/build/main/lib/map-options.js
/**
 * Only partial compatibility is achieved here. Notes:
 *
 * - Cannot do percentage-based dimension changes using our streaming strategy
 * - Same for percentage-based crops and aspect-ratio-based crops
 */
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const accepts_1 = __importDefault(require("accepts"));
const background_flatten_1 = __importDefault(require("./mappers/background-flatten"));
const blur_1 = __importDefault(require("./mappers/blur"));
const extend_1 = __importDefault(require("./mappers/extend"));
const extract_crop_1 = __importDefault(require("./mappers/extract-crop"));
const orient_1 = __importDefault(require("./mappers/orient"));
const resize_1 = __importDefault(require("./mappers/resize"));
const resize_canvas_1 = __importDefault(require("./mappers/resize-canvas"));
const unsupported_1 = __importDefault(require("./mappers/unsupported"));
const mappers = [
    ['width', resize_1.default],
    ['height', resize_1.default],
    ['dpr', resize_1.default],
    ['fit', resize_1.default],
    ['disable', resize_1.default],
    ['enable', resize_1.default],
    ['resize-filter', resize_1.default],
    ['crop', extract_crop_1.default],
    ['canvas', resize_canvas_1.default],
    ['pad', extend_1.default],
    ['orient', orient_1.default],
    ['bg-color', background_flatten_1.default],
    ['blur', blur_1.default],
    ['brightness', unsupported_1.default('brightness', 'absolute brightness adjustment')],
    ['contrast', unsupported_1.default('contrast', 'absolute contrast adjustment')],
    ['saturation', unsupported_1.default('saturation', 'absolute saturation adjustment')],
    ['sharpen', unsupported_1.default('sharpen', 'unsharp mask')],
    ['trim', unsupported_1.default('trim', 'relative trimming from all four sides')],
];
const formatters = {
    gif: unsupported_1.default('format', 'GIF output unsupported by node-hastily'),
    jpg: (transform, params) => transform.jpeg({ quality: params.quality }),
    pjpg: (transform, params) => transform.jpeg({ quality: params.quality, progressive: true }),
    png: (transform, params) => transform.png({ quality: params.quality }),
    png8: (transform, params) => transform.png({ palette: true, quality: params.quality }),
    webp: (transform, params) => transform.webp({ quality: params.quality }),
    webpll: (transform, params) => transform.webp({ quality: params.quality, lossless: true }),
    webply: (transform, params) => transform.webp({ quality: params.quality }),
};
/**
 * @hidden
 */
function supportsWebP(req) {
    const types = accepts_1.default(req).types();
    if (!types || (typeof types === 'string' && types !== 'image/webp')) {
        return false;
    }
    return types.includes('image/webp');
}
/**
 * @hidden
 */
function optoToSharp(params, sharpStream) {
    const applied = new Set();
    const { req, res, log } = params;
    const { query } = req;
    let transform = sharpStream;
    for (const [name, mapper] of mappers) {
        if (query.hasOwnProperty(name) && !applied.has(mapper)) {
            log.debug('running mapper for %s', name);
            const out = mapper(transform, params);
            if (out) {
                applied.add(mapper);
                transform = out;
            }
        }
    }
    if (query.auto === 'webp' && supportsWebP(req)) {
        log.debug('returning webp');
        res.type('image/webp');
        return formatters.webp(transform, params);
    }
    const arguedFormat = query.format;
    const tryJpeg = (xform, { quality }) => {
        if (applied.size === 0) {
            log.debug('no mappers or formatters applied, doing nothing');
            return false;
        }
        return xform.jpeg({
            force: false,
            quality,
        });
    };
    if (!arguedFormat) {
        log.debug('no format argument, returning jpeg or whatever');
        res.type('image/jpeg');
        return tryJpeg(transform, params);
    }
    const mapFormat = formatters[arguedFormat];
    if (typeof mapFormat !== 'function') {
        log.debug('bad format argument %s, returning jpeg', arguedFormat);
        return tryJpeg(transform, params);
    }
    log.debug('attempting "%s" transform', arguedFormat);
    res.type(`image/${arguedFormat}`);
    return mapFormat(transform, params);
}
exports.default = optoToSharp;
